ArmPkg/Drivers/SmmuDxe/IoMmu.c

```diff
@@ -0,0 +1,523 @@
+/** @file IoMmu.c
+
+    This file contains functions for the IoMmu protocol.
+
+    Copyright (c) Microsoft Corporation. All rights reserved.
+    SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#include <Uefi.h>
+#include <Library/ArmLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/DebugLib.h>
+#include <Library/IoLib.h>
+#include <Library/MemoryAllocationLib.h>
+#include <Library/TimerLib.h>
+#include <Library/UefiBootServicesTableLib.h>
+#include <Library/UefiDriverEntryPoint.h>
+#include <Protocol/IoMmu.h>
+#include "IoMmu.h"
+#include "SmmuV3Registers.h"
+
+#define PAGE_TABLE_DEPTH  4                             // Number of levels in the page table
+#define PAGE_TABLE_READ_WRITE_FROM_IOMMU_ACCESS(IOMMU_ACCESS)  (IOMMU_ACCESS << 6)
+#define PAGE_TABLE_READ_BIT         (0x1 << 6)
+#define PAGE_TABLE_WRITE_BIT        (0x1 << 7)
+#define PAGE_TABLE_ENTRY_VALID_BIT  0x1
+#define PAGE_TABLE_BLOCK_OFFSET     0xFFF
+#define PAGE_TABLE_ACCESS_FLAG      (0x1 << 10)
+#define PAGE_TABLE_DESCRIPTOR       (0x1 << 1)
+#define PAGE_TABLE_INDEX(VA, LEVEL)  (((VA) >> (12 + (9 * (PAGE_TABLE_DEPTH - 1 - (LEVEL))))) & 0x1FF)
+
+EDKII_IOMMU_PROTOCOL  SmmuIoMmu = {
+  EDKII_IOMMU_PROTOCOL_REVISION,
+  IoMmuSetAttribute,
+  IoMmuMap,
+  IoMmuUnmap,
+  IoMmuAllocateBuffer,
+  IoMmuFreeBuffer,
+};
+
+typedef struct IOMMU_MAP_INFO {
+  UINTN     NumberOfBytes;
+  UINT64    VA;
+  UINT64    PA;
+} IOMMU_MAP_INFO;
+
+/**
+  Update the flags of a page table entry.
+
+  @param [in]  Table         Pointer to the page table.
+  @param [in]  SetFlagsOnly  Boolean to indicate if only flags should be set.
+  @param [in]  Flags         Flags to set or clear.
+  @param [in]  Index         Index of the entry to update.
+
+  @retval EFI_SUCCESS        Success.

```

Comment:
Double check these. For example, the `EFI_INVALID_PARAMETER` parameter is missing.

Reply:
fixed, here and throughout.